### PR TITLE
ExistsIn() should allow composite foreign keys having NULL allowed in the schema

### PR DESCRIPTION
**Pull request by ionas**
Ref: https://github.com/cakephp/cakephp/issues/8749#issuecomment-222028261

### CODE REVIEW

File: src/ORM/Rule/ExistsIn.php [hunk 1]
```diff
@@ -40,14 +40,24 @@ class ExistsIn
...
      * @param string|array $fields The field or fields to check existence as primary key.
      * @param object|string $repository The repository where the field will be looked for,
      * or the association name for the repository.
      */
-    public function __construct($fields, $repository)
+    public function __construct($fields, $repository, $setupOptions = array())
     {
+        $this->_setupOptions = [
+                'allowPartialSchemaNulls' => true,
+            ] + $setupOptions;
```
Comment (markstory) [inline src/ORM/Rule/ExistsIn.php]: I think this needs to be be `$options + $defaults`. Right now, it is impossible to make `allowPartialSchemaNulls = false`.

**Comment by markstory** [comment]
This looks related to #8621 which added the ability to be stricter than SQL `UNIQUE` keys can be.

File: src/ORM/Rule/ExistsIn.php (hunk 1 above)
Comment (ionas) [inline src/ORM/Rule/ExistsIn.php]: Should be fixed now.

**Comment by ionas** [comment]
Use like:
```php
        $rules->add($rules->existsIn(['parent_id', 'instance_id'], 'ParentNodes', ['allowSqlNulls' => true]));
```

File: src/ORM/Rule/ExistsIn.php [hunk 2]
```diff
@@ -131,4 +145,22 @@ protected function _fieldsAreNull($entity, $source)
         }
         return $nulls === count($this->_fields);
     }
+
+    /**
+     * Check wether there are partial, nullable, nulls.
```
Comment (markstory) [inline src/ORM/Rule/ExistsIn.php]: Typo in whether.

File: src/ORM/Rule/ExistsIn.php [hunk 3]
```diff
@@ -40,14 +40,24 @@ class ExistsIn
...
+
+    /**
      * Constructor.
      *
      * @param string|array $fields The field or fields to check existence as primary key.
      * @param object|string $repository The repository where the field will be looked for,
      * or the association name for the repository.
      */
-    public function __construct($fields, $repository)
+    public function __construct($fields, $repository, array $constructOptions = Array())
```
Comment (markstory) [inline src/ORM/Rule/ExistsIn.php]: `Array()` should be `[]`

File: src/ORM/Rule/ExistsIn.php [hunk 4]
```diff
@@ -40,14 +40,24 @@ class ExistsIn
...
     {
+        $constructOptions += ['allowSqlNulls' => false];
+        $this->_constructOptions = $constructOptions;
```
Comment (markstory) [inline src/ORM/Rule/ExistsIn.php]: `$this->_options`.

**Comment by markstory** [comment]
This should probably target `3.next`

File: src/ORM/Rule/ExistsIn.php [hunk 5]
```diff
@@ -98,7 +108,11 @@ public function __invoke(EntityInterface $entity, array $options)
             return true;
         }
 
-        if ($this->_fieldsAreNull($entity, $source)) {
+        if ($this->_constructOptions['allowSqlNulls'] === true
+            && $this->_checkPartialSchemaNulls($entity, $source) === true
+        ) {
+            return true;
+        } elseif ($this->_fieldsAreNull($entity, $source)) {
```
Comment (dereuromark) [inline src/ORM/Rule/ExistsIn.php]: no else required when returning early.

File: src/ORM/Rule/ExistsIn.php (hunk 4 above)
Comment (ionas) [inline src/ORM/Rule/ExistsIn.php]: Changing. Reason was that invoke and construct can get options. This would be the difference.

**Comment by ionas** [comment]
> This should probably target `3.next`

how?

**Comment by ionas** [comment]
@markstory I have added tests. Is this what you had in mind?

File: tests/Fixture/SiteCategoriesFixture.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+     * fields property
+     *
+     * @var array
+     */
+    public $fields = [
+        'id' => ['type' => 'integer'],
+        'parent_id' => ['type' => 'integer', 'null' => true],
+        'site_id' => ['type' => 'integer', 'null' => false],
+        'name' => ['type' => 'string', 'null' => false],
+        '_constraints' => ['primary' => ['type' => 'primary', 'columns' => ['site_id', 'id']]]
```
Comment (ADmad) [inline tests/Fixture/SiteCategoriesFixture.php]: We already have a similar `SiteArticlesFixture` with composite key. You can just use that instead of creating a new fixture.

**Comment by ADmad** [comment]
> how? Edit: can I change the target?

You can't change the target branch for a PR. You will need to open an new PR targeting 3.next.

File: tests/Fixture/SiteCategoriesFixture.php (hunk 6 above)
Comment (ionas) [inline tests/Fixture/SiteCategoriesFixture.php]: Before creating a fixture I went through all of them. I wanted to mimic self-join and there is no fixture with self-join and composite primary key, is there?

**Comment by ionas** [comment]
Awesome gh misses tiny things. @Admad once you are happy and it is ready for merge I will close and re-open.

File: src/ORM/Rule/ExistsIn.php [hunk 7]
```diff
@@ -98,6 +109,11 @@ public function __invoke(EntityInterface $entity, array $options)
             return true;
         }
 
+        if ($this->_options['allowSqlNulls'] === true
+            && $this->_checkPartialSchemaNulls($entity, $source) === true
+        ) {
+            return true;
```
Comment (chinpei215) [inline src/ORM/Rule/ExistsIn.php]: I cannot see what you are trying to do in this PR. This code would return `true` against key pairs such as the following:
```php
[
    'parent_id' => null,
    'site_id' => 99999, -- Non existent
]
```

But seems to be intended. Because `_fieldsAreNull()` is enough to return `true` against key pairs such as the following:
```php
[
    'parent_id' => null,
    'site_id' => 1, -- Existent
]
```

So how does this feature benefit you? I think you might want to explain it in more detail.

File: tests/Fixture/SiteCategoriesFixture.php (hunk 6 above)
Comment (chinpei215) [inline tests/Fixture/SiteCategoriesFixture.php]: I agree with @ADmad, as this PR doesn't seem to be related with self-join.

File: src/ORM/Rule/ExistsIn.php (hunk 7 above)
Comment (ionas) [inline src/ORM/Rule/ExistsIn.php]: Before I (try to) swap the fixtures, take a look at the test, fixture and example use case and it should be (almost) obvious why.

For an ExistsIn() constraint where the referenced composite key may be partially null if just one of the keys that MAY BE null IS NULL then the whole rule should return true. If you need to be more specific you want to:
- Setup additional ExistsIn() constraints on smaller parts (like only site_id),...
- ... or set all fields to be NOT NULL.

So yes, the intended behavior is such that the top rule will pass because it cannot exist anyway because parent_id is null. If you don't want to check if for instance a Category, given a Site exists in a ParentCategory (given the same Site) but only want to check if a Category references a correct and existing Site then that's what you do:

```php
        $rules->add($rules->existsIn(['site_id'], 'Sites')); // Does this SiteCategory reference an existing Site.
        $rules->add($rules->existsIn(['parent_id', 'site_id'], 'ParentNodes', ['allowSqlNulls' => true])); // Does this SiteCategory reference and existing Parent SiteCategory or none at all (parent_id = null)
```

Comment (ionas) [inline src/ORM/Rule/ExistsIn.php]: @markstory should the option possibly be named trueOnPartialNulls?

Comment (chinpei215) [inline src/ORM/Rule/ExistsIn.php]: > // Does this SiteCategory reference an existing Site.
> $rules->add($rules->existsIn(['site_id'], 'Sites'));
> // Does this SiteCategory reference and existing Parent SiteCategory or none at all (parent_id = null)
> $rules->add($rules->existsIn(['parent_id', 'site_id'], 'ParentNodes', ['allowSqlNulls' => true]));

@ionas Sorry, i cannot understand yet :cry: How does the above behave differently from the following rules?
```php
$rules->add($rules->existsIn(['site_id'], 'Sites'));
$rules->add($rules->existsIn(['parent_id'], 'ParentNodes'));
```

Comment (ionas) [inline src/ORM/Rule/ExistsIn.php]: First of all - Are you at CakeFest :)?

While I could get away with that regarding Tree Behavior because it is hard to teach it composite foreign keys, the basic premise here is that the real foreign key is not just parent_id but [site_id, parent_id], e.g. site_id is both: a foreignKey and part of the composite primary key.

In other words:
In case you reference a foreign table where one field can be null and IS null, the foreign key reference is actually optional unless you specify all foreign key parts and then it HAS to exist.

Does the current `__fieldsAreNull()` check make sense to you?

File: tests/Fixture/SiteCategoriesFixture.php (hunk 6 above)
Comment (ionas) [inline tests/Fixture/SiteCategoriesFixture.php]: Gave it another try: Doesn't work because AuthorFixture does not have a site_id so the feature cannot be tested by Authors + SiteArticles.

File: src/ORM/Rule/ExistsIn.php (hunk 7 above)
Comment (chinpei215) [inline src/ORM/Rule/ExistsIn.php]: No. But I want to come there when I become able to speak/listen English.

And I understand what you want. As I always use a surrogate key, I didn't know how NULL values in a composite foreign key are handled in SQL. I thought that MySQL/PostgreSQL rejects to insert a composite key like the above ( `[ parent_id => null, site_id => 99999 ]` ), 
but they actually allow to insert any values if it contains partially NULL.

The current `_fieldsAreNull` behavior seems incompatible with MySQL and PostgreSQL at least, when a composite foreign key allows NULL values. Personally I think you can fix the current behavior instead of adding a new option. (Because incompatible with common databases)

Comment (ionas) [inline src/ORM/Rule/ExistsIn.php]: For the matter at hand:
I did at first only fix the issue (see the commit log). Problem: people expect things to work like this now and it is not backwards compatible. Because of that I have added the option. For 4.x or 5.x we can switch the flag to true by default and/or introduce a flag to re-enable the old behavior.

Personal Note: In case you will understand what others are saying you should join. We had some people from Japan even giving some talks. Love to see you next time!

File: tests/Fixture/SiteCategoriesFixture.php (hunk 6 above)
Comment (chinpei215) [inline tests/Fixture/SiteCategoriesFixture.php]: @ionas I took a look at existing fixtures. `SiteAuthorsFixture` seems suitable to me. Doesn't it work?

File: src/ORM/Rule/ExistsIn.php (hunk 7 above)
Comment (chinpei215) [inline src/ORM/Rule/ExistsIn.php]: @ionas Indeed you did it in the first commit. I simply think that if the current behavior is a bug, we can fix it in 3.x. If it's a feature then 4.x might be suitable as you said.

At the moment, I cannot understand what they are telling me at all, if they are speaking English, not writing :stuck_out_tongue: Furthermore I have never been abroad. But I am planning to join next CakeFest. I hope to see you next time :cake:

File: tests/Fixture/SiteCategoriesFixture.php (hunk 6 above)
Comment (ionas) [inline tests/Fixture/SiteCategoriesFixture.php]: @chinpei215 Thank you I did somehow overlook SiteAuthorsFixture. It does work now.

File: src/ORM/Rule/ExistsIn.php [hunk 8]
```diff
@@ -40,14 +40,25 @@ class ExistsIn
...
+     */
+    protected $_options = [];
+
+    /**
      * Constructor.
      *
      * @param string|array $fields The field or fields to check existence as primary key.
      * @param object|string $repository The repository where the field will be looked for,
      * or the association name for the repository.
+     * @param array $options The options that modify the rules behavior.
```
Comment (ADmad) [inline src/ORM/Rule/ExistsIn.php]: Description of valid options should be provided here. Just by it's name I have no idea what `passingOnPartialNulls` does, also it's a bit too wordy imo.

File: src/ORM/Rule/ExistsIn.php [hunk 9]
```diff
@@ -98,6 +109,11 @@ public function __invoke(EntityInterface $entity, array $options)
             return true;
         }
 
+        if ($this->_options['passingOnPartialNulls'] === true
```
Comment (ADmad) [inline src/ORM/Rule/ExistsIn.php]: The option is a boolean and return value of the method is also boolean. So can just do:
`if ($this->_options['passingOnPartialNulls'] && $this->_checkPartialSchemaNulls($entity, $source))`

File: src/ORM/RulesChecker.php [hunk 10]
```diff
@@ -71,14 +71,25 @@ public function isUnique(array $fields, $message = null)
      * $rules->add($rules->existsIn('site_id', new SitesTable(), 'Invalid Site'));
      * ```
      *
+     * Available $options are error 'message' and 'allowSqlNulls' flag.
```
Comment (ADmad) [inline src/ORM/RulesChecker.php]: Shouldn't `allowSqlNulls` be `passingOnPartialNulls`?

File: src/ORM/RulesChecker.php [hunk 11]
```diff
@@ -71,14 +71,25 @@ public function isUnique(array $fields, $message = null)
...
     {
+        $message = null;
+        if (is_string($options)) {
+            $message = $options;
+            $options = null;
+        } elseif (is_array($options)) {
+            $options += ['message' => ''];
+            $message = $options['message'];
+            unset($options['message']);
+        }
```
Comment (ADmad) [inline src/ORM/RulesChecker.php]: Following would be much more readable:

```php
if (is_string($options)) {
    $options = ['message' => $options];
}

$options = (array)$options + ['message' => null];
$message = $options['message'];
unset($options['message']);
```

Would also remove the need to typecast $options to array on line 102.

File: src/ORM/Rule/ExistsIn.php (hunk 9 above)
Comment (ionas) [inline src/ORM/Rule/ExistsIn.php]: I changed it. However I don't understand why it is wrong to be explicit about types `in general`.

File: src/ORM/Rule/ExistsIn.php (hunk 8 above)
Comment (ionas) [inline src/ORM/Rule/ExistsIn.php]: I did shorten it. I hoped that the flag was speaking for itself. If you take a look at the commits you will see that I was really looking for a good name.

**Comment by ionas** [comment]
Is this now ready for merge, so that I can re-open against 3.next?

File: src/ORM/Rule/ExistsIn.php [hunk 12]
```diff
@@ -40,14 +40,28 @@ class ExistsIn
...
+     *
+     * @var array
+     */
+    protected $_options = [];
+
+    /**
      * Constructor.
      *
+     * Available option for $options is 'partialNullsPass' flag. 
+     * Set to true to allow passing of composite foreign keys where one or multiple keys are null.
```
Comment (markstory) [inline src/ORM/Rule/ExistsIn.php]: 'Set to true to accept composite foreign keys where one or more nullable columns are null.'

File: src/ORM/Rule/ExistsIn.php [hunk 13]
```diff
@@ -40,14 +40,28 @@ class ExistsIn
...
+     *
      * @param string|array $fields The field or fields to check existence as primary key.
      * @param object|string $repository The repository where the field will be looked for,
      * or the association name for the repository.
+     * @param array $options The options that modify the rules behavior.
      */
-    public function __construct($fields, $repository)
+    public function __construct($fields, $repository, array $options = [])
     {
+        $options += ['partialNullsPass' => false];
```
Comment (markstory) [inline src/ORM/Rule/ExistsIn.php]: I'm wondering if we've found a good name for this:

* `acceptNulls`
* `allowNullValues`

Comment (ionas) [inline src/ORM/Rule/ExistsIn.php]: `allowPartialNulls`?

Comment (markstory) [inline src/ORM/Rule/ExistsIn.php]: `allowPartialNulls` works for me.

**Comment by ionas** [comment]
Closing and re-targeting against 3.next.
